File: P1/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

bool checksum(long n, int digit);
void checkserver(long n, int digit);

int main(void)
{
    // user inputs card number
    long n = get_long("Number: ");

    // calculate how many digits the number has
    int digit = (int) log10(n) + 1;

    // checksum return true, card is valid
    if (checksum(n, digit))
    {
        // check card server
        checkserver(n, digit);
    }
    else
    {
        // checksum return false, card is invalid
        printf("INVALID\n");
    }

    return 0;
}

// checksum: Luhn’s Algorithm
bool checksum(long n, int digit)
{
    int sum = 0;

    // From second-to-last digit to get every other digit
    for (int i = 1; i < digit; i = i + 2)
    {
        int x = (n / (long) pow(10, i)) % 10;
        // Multiply by 2
        int y = x * 2;
        // add those products’ digits (NOT the products themselves) together
        sum += y / 10 + y % 10;
    }

    // Add the sum of the digits that weren’t multiplied by 2
    for (int i = 0; i < digit; i = i + 2)
    {
        sum += (n / (long) pow(10, i)) % 10;
    }

    // If the total’s last digit is 0, the number is valid.
    return sum % 10 == 0;
}

// check card server
void checkserver(long n, int digit)
{
    int x = n / (long) pow(10, digit - 2);

    // American Express uses 15-digit numbers
    if (digit == 15)
    {
        // All American Express numbers start with 34 or 37
        if (x == 34 || x == 37)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        // Visa uses 13- and 16-digit numbers
        if (digit == 13)
        {
            // All Visa numbers start with 4
            if (x / 10 == 4)
            {
                printf("VISA\n");
            }
        }
        else
        {
            if (digit == 16)
            {
                // MasterCard uses 16-digit numbers
                // Most MasterCard numbers start with 51, 52, 53, 54, or 55
                if (x == 51 || x == 52 || x == 53 || x == 54 || x == 55)
                {
                    printf("MASTERCARD\n");
                }
                else
                {
                    // Visa also uses 16-digit numbers
                    // All Visa numbers start with 4
                    if (x / 10 == 4)
                    {
                        printf("VISA\n");
                    }
                    else
                    {
                        printf("INVALID\n");
                    }
                }
            }
            else
            {
                printf("INVALID\n");
            }
        }
    }
}

File: P6/credit.py
from cs50 import get_int


def checksum(n):
    # check if the card is valid with Luhn’s Algorithm
    sum = 0
    # get an array of reversed digits
    reverse = [int(d) for d in str(n)][::-1]

    for i, digit in enumerate(reverse):
        if i % 2 == 1:
            product = digit * 2
            sum += product // 10 + product % 10
        else:
            sum += digit

    return sum % 10 == 0


def checkserver(n):
    # check the card server
    digits = len(str(n))
    prefix = int(str(n)[:2])

    # AMEX
    if digits == 15 and prefix in {34, 37}:
        return "AMEX"
    # MASTERCARD
    elif digits == 16 and 51 <= prefix <= 55:
        return "MASTERCARD"
    # VISA
    elif digits in {13, 16} and 40 <= prefix <= 49:
        return "VISA"
    # INVALID
    return "INVALID"


def main():
    # prompt for a card number
    n = get_int("Number: ")

    # print the result
    if checksum(n):
        print(checkserver(n))
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
